### PR TITLE
Initial commit

### LINKED ISSUE #1: Add secret store
Secret story initially has been created in the scope of ticket https://github.com/exasol/data-science-sandbox/issues/1 

The current ticket requests to move the implementation to this dedicated repository

### PR DESCRIPTION
**Pull request by ckunki**
closes #1

### CODE REVIEW

File: .github/actions/prepare_poetry_env/action.yml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  steps:
+    - uses: actions/setup-python@v2
+      with:
+        python-version: ${{ inputs.python-version }}
+    - uses: abatilo/actions-poetry@v2
+      with:
+        poetry-version: 1.4.0
+    - name: Poetry install
+      run: poetry install
+      shell: bash
```
Comment (kaklakariada) [inline .github/actions/prepare_poetry_env/action.yml:19]:
```suggestion
```
Not necessary

File: .github/actions/prepare_poetry_env/action.yml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  python-version:
+    description: 'The Python version to use'
+    required: true
+    default: '3.8'
+runs:
+  using: "composite"
+  steps:
+    - uses: actions/setup-python@v2
+      with:
+        python-version: ${{ inputs.python-version }}
```
Comment (kaklakariada) [inline .github/actions/prepare_poetry_env/action.yml:13]: Configure caching for python packages

File: .github/workflows/ci_build.yaml [hunk 3]
```diff
@@ -0,0 +1,20 @@
+name: CI Build
+
+on:
+  push:
+    branches-ignore:
+      - "main"
+
+jobs:
+  run_unit_tests:
+    runs-on: ubuntu-latest
```
Comment (kaklakariada) [inline .github/workflows/ci_build.yaml:10]:
```suggestion
    runs-on: ubuntu-latest
    concurrency:
      group: ${{ github.workflow }}-${{ github.ref }}
      cancel-in-progress: true
```

File: .github/workflows/ci_build.yaml [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+    steps:
+     - uses: actions/checkout@v3
```
Comment (kaklakariada) [inline .github/workflows/ci_build.yaml:13]:
```suggestion
     - uses: actions/checkout@v4
```

File: notebook_connector/secret_store.py
Comment (kaklakariada) [inline notebook_connector/secret_store.py:1]: I did not review this, assuming it was already reviewed.

File: notebook_connector/poetry.lock
Comment (kaklakariada) [inline notebook_connector/poetry.lock:1]: There are two `poetry.lock` files. Do we really need both?

File: notebook_connector/secret_store.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+_logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class Table:
+    name: str
+    columns: List[str]
+
+
+SECRETS_TABLE = Table("secrets", ["user", "password"])
```
Comment (tkilias) [inline notebook_connector/secret_store.py:19]: I think, we have not only have user and password in the worst case. Maybe, we should define a Secret as a Key, Value pair and you can use two key value pairs for user and password,

(Can be done in a later PR)

File: notebook_connector/secret_store.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+    user: str
+    password: str
+
+
+class InvalidPassword(Exception):
+    """Signal potentially incorrect master password."""
+
+
+class Secrets:
+    def __init__(self, db_file: str, master_password: str) -> None:
```
Comment (tkilias) [inline notebook_connector/secret_store.py:34]:
```suggestion
    def __init__(self, db_file: Path, master_password: str) -> None:
```

can be done in a later PR

File: notebook_connector/secret_store.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+        self._con = None
+
+    def close(self) -> None:
+        if self._con is not None:
+            self._con.close()
+            self._con = None
+
+    def connection(self) -> sqlcipher.Connection:
+        if self._con is None:
+            db_file_found = os.path.exists(self.db_file)
```
Comment (tkilias) [inline notebook_connector/secret_store.py:46]: I prefer pathlib.Path.exists, the os package is not recommended anymore (can be done in a later PR)

File: test/test_secret_store.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    assert secrets.config("key") == other
+
+
+def test_wrong_password(sample_file):
+    secrets = Secrets(sample_file, "correct password")
+    secrets.save("key", Credentials("usr", "pass")).close()
+    invalid = Secrets(sample_file, "wrong password")
+    with pytest.raises(InvalidPassword) as ex:
+        invalid.credentials("key")
+    assert "master password is incorrect" in str(ex.value)
```
Comment (tkilias) [inline test/test_secret_store.py:56]: maybe add a test where you try to read a sqlcipher db with sqlite, to see if sqlcipher actually encrypts it.

(Can be done in a later PR)

File: pyproject.toml [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+
+[build-system]
+requires = ["poetry_core>=1.0.0"]
+build-backend = "poetry.core.masonry.api"
+
+
+[tool.poetry.dev-dependencies]
+pytest = "^7.1.1"
+pytest-mock = "^3.7.0"
```
Comment (tkilias) [inline pyproject.toml:24]: add python toolbox and nox (can be done in a later PR)

**Review by kaklakariada — APPROVED** [review]

**Pull request merged**